### PR TITLE
fix(quantization): bound NVFP4 loader scale memory on Karmic

### PR DESCRIPTION
**Pull request by voipmonitor**
## Purpose

Reduce temporary VRAM during NVFP4 weight loading, without changing inference
precision or kernels. GLM Spark TP2 can otherwise fail before KV-cache allocation.

- Reduce input scales to a scalar before converting to FP32.
- Convert Marlin scale banks one expert at a time into their final output.
- Reduce the MTP vocabulary head in bounded FP32 chunks instead of copying the
  complete BF16 head into FP32.

Finite checkpoint scale factors, packed weights and input ownership are
preserved. NaN Marlin scales raise explicitly; the head retains its existing
nonfinite-value handling.

## Attribution and related work

This is a Karmic port of Elevatormusic's upstream
[vLLM #55103](https://github.com/vllm-project/vllm/pull/55103), with the original
author, coauthor and commit references retained. It also includes the bounded
expert-bank and MTP-head reductions. No duplicate upstream PR is proposed.
LIL #739 targets JJ and still materializes a complete FP32 head; this port
targets `dev/karmic-kraken` and bounds that allocation.

## Validation

Status: qualified component memory/bit-parity checks and composed TP2 serving.
CUDA 13.4.1, PyTorch 2.14, RTX PRO 6000 Workstation at stock clocks:

- Scalar reduction: 26 tests pass. A 72 MiB BF16 input needs 2,560 bytes of
  auxiliary allocation, versus 1,243,946,496 bytes in the mask/gather control.
- Expert banks: 35 tests pass, including bitwise scale-byte parity. A 27 MiB
  output needs 28,775,936 auxiliary bytes versus 75,503,616 in the control.
- Draft head: nine tests pass, including all BF16 bit patterns and packed-byte
  parity. A 32 MiB BF16 test needs 16,781,824 auxiliary bytes; the full-copy
  control needs 134,217,728. Negative allocation controls fail as expected.

The production files are byte-identical to those in the qualified KK TP2
source tree. Whole-model serving requires the accompanying workspace and
capacity PRs; this loading change alone is not a complete TP2 recipe.

Tests: `tests/kernels/quantization/test_nvfp4_marlin_scale_factor.py` and the
draft-head cases in `tests/models/test_glm5next_model.py`.
OpenAI Codex assisted with porting and validation; maintainer review is required.

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved NVFP4 quantization reliability for empty inputs, non-finite values, varied data types, and zero-valued tensors.
  - Preserved accurate scaling and packed weights during GLM-5.3 draft-head quantization.

- **Performance**
  - Reduced temporary memory usage during mixture-of-experts and draft-head scale preparation, including for large model weights.

- **Tests**
  - Added extensive coverage for NVFP4 scaling, quantization accuracy, memory limits, and edge-case handling.

### CODE REVIEW

**Comment by voipmonitor** [comment]
@coderabbitai full review

Please review the complete proposed change, including ownership, graph lifetime and the recorded validation. Serving changes remain for maintainer review; integration qualification is not approval.

File: tests/kernels/quantization/test_nvfp4_marlin_scale_factor.py [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+
+def test_scale_factor_rescales_into_range():
+    scales = torch.full((32, 64), 2**-10, dtype=torch.bfloat16)
+    sf = _nvfp4_compute_scale_factor(scales)
+    rescaled = scales.float().max() * (2**7) * sf
+    assert 2.0 <= rescaled < UPPER
+
+
+def test_scale_factor_empty_tensor():
+    assert _nvfp4_compute_scale_factor(torch.empty(0, dtype=torch.bfloat16)) == 1.0
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply tests/kernels/quantization/test_nvfp4_marlin_scale_factor.py:87]: Checked the submitted head d6a88c611: `awk 'length($0)>88 {print NR, length($0)}' tests/kernels/quantization/test_nvfp4_marlin_scale_factor.py` produces no output. Both referenced lines already satisfy the 88-column limit; no formatting change is needed.

File: vllm/model_executor/layers/quantization/utils/marlin_utils_fp4.py [hunk 2]
```diff
@@ -47,14 +47,17 @@ def _nvfp4_compute_scale_factor(
...
-    if nonzero_mask.any():
-        max_val = ws_float[nonzero_mask].max()
-        if max_val < 448 * (2**7):
-            sf = (448 * (2**7) / max_val).log2().floor().exp2()
-            return sf.item()
+    # Reduce before conversion to keep temporary storage independent of the
+    # number of experts. Power-of-two multiplication preserves the maximum.
+    max_val = marlin_scales.max().float() * (2**7)
+    if torch.isnan(max_val):
+        raise ValueError("NVFP4 Marlin scales contain NaN")
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply vllm/model_executor/layers/quantization/utils/marlin_utils_fp4.py:68]: Documented the arguments, returned factor and NaN ValueError in a5bab53f7bf. The FP16 bypass and empty/non-positive inputs are explicit. A Python AST comparison with docstrings removed is identical before/after; Ruff and all applicable repository pre-commit hooks pass. No arithmetic or model evaluation changed.
